### PR TITLE
[WIP] Decompiler Improvements

### PR DESCRIPTION
**Pull request by ritzdorf**
* Adding additional EVM instructions to allow better decompilation (Revert, Returndatasize, Returndatacopy, Staticcall, Invalid)
* Hotfixing a bug that causes the intermediate representation to fail when a conditional jump is leading to two stack merges 
* Hotfixing another bug that causes the decompilation to crash in case of unbalanced stacks. Independently verified the jumps and the stack growth. They seem correct. A proper fix requires an slight adaptation in the abstraction.

Here is a breaking example for the first issue:
https://gist.github.com/ritzdorf/7aaa2f59a3be273d1aab9ee4d929219b
This should work now with the hotfix.

Generally, a lot more testing is needed.

### CODE REVIEW

File: src/main/java/ch/securify/decompiler/ControlFlowDetector.java [hunk 1]
```diff
@@ -221,6 +221,12 @@ else if (opcode == OpCodes.RETURN) {
 				richBranches.put(pc, BRANCH_EXIT);
 				return;
 			}
+			else if (opcode == OpCodes.REVERT) {
```
Comment (jchittoda) [inline src/main/java/ch/securify/decompiler/ControlFlowDetector.java]: As REVERT opcode is added in ControlFlowDetector. Do we need to add StaticCall also here, in this class?

File: src/main/java/ch/securify/decompiler/evm/OpCodes.java [hunk 2]
```diff
@@ -166,9 +166,10 @@ public static int isSwap(int opcode) {
 	public static final int CALLCODE = 0xf2;
 	public static final int RETURN = 0xf3;
 	public static final int DELEGATECALL = 0xf4;
+	public static final int STATICCALL = 0xf5;
```
Comment (jchittoda) [inline src/main/java/ch/securify/decompiler/evm/OpCodes.java]: As per this yellow paper https://ethereum.github.io/yellowpaper/paper.pdf, STATICCALL code is 0xfa. 
Is there any other place to check for the latest opcodes of the instructions?

File: src/main/java/ch/securify/decompiler/ControlFlowDetector.java (hunk 1 above)
Comment (ritzdorf) [reply src/main/java/ch/securify/decompiler/ControlFlowDetector.java]: This is computing the internal control flow graph. So I don't think it is needed here. But we still need to tag the output of StaticCall as uncontrolled. For this I need to synch with @ptsankov on the correct memory abstraction.

**Comment by hiqua** [comment]
@ritzdorf are there other changes you want to add to this PR or is it no longer a WIP?

**Comment by ritzdorf** [comment]
Will finish it up today.

**Comment by hiqua** [comment]
Ok just remove the [WIP] from the title when you think it should be merged.

**Comment by ritzdorf** [comment]
@hiqua Ready for merge from my side.

**Comment by ritzdorf** [comment]
Found another case that is not covered by my fix. Will give a time estimate as soon as I can.

**Comment by hiqua** [comment]
@ritzdorf can you expand on this case here directly?

**Comment by hiqua** [comment]
Closing as discussed with @ritzdorf.

**Pull request closed without merging**